# Python 101:  Working with Data

You are reading a Jupyter Notebook.  Hopefully you have already worked through the Jupyter Notebooks tutorial and you have the nbextensions installed, and the outline view and spell checker enabled.  In addition, be sure to toggle line numbers on from the `View` menu above.  Remember, you can make the text bigger with `ctrl-+`.  (https://realpython.com/jupyter-notebook-introduction/).  For now we will work with two kinds of cells.  The cell you are reading now is called a Markdown cell.  Look at the second row of the menu.  On the right you will see `Markdown` in a drop down list box.  This tells you you are in a markdown cell. 

For more on Jupyter Notebooks go here https://www.dataquest.io/blog/jupyter-notebook-tips-tricks-shortcuts/

## Python codes

[100 Basic Python Codes Cookbook](https://newdigitals.org/2024/02/24/100-basic-python-codes/)

## Hello World

It is traditional when using a new piece of software that you write and execute a hello world program.  This shows you that the software is working on your computer and that you can access it.  Go ahead and click on the cell below this one so it is outlined in green and then type Ctrl-Enter (Note this is shorthand for saying press down on the Ctrl key and the Enter key at the same time).  

In [1]:
# Hello World Program
print('hello world')

hello world


Line 1 is a comment and ignored by Python.  Line 2 calls the built-in `print` function in Python and has it display `hello world` on your screen.
   

# Data and Variables

In programming we have **data** and **statements** that manipulate data.  To work with data we must first get access to the data we want to process.  We have many options, for example: 

1. We can put the data in our program in what are called **literals**.  
2. We can enter the data in our program using the **input** function.    
3. We can **read** data from a file.
4. We can **capture** data from sensors built into hardware.
5. we can **scrape** data from web pages. 

Before a program can use data it must be put into our computer's memory.  Since computer memory is just zeros (physical off) and ones (physical on), data is encoded in memory according to its **type**, that is, the pattern of zeros and ones that represent the data.  For example, **boolean data** (**True** and **False**) are encoded differently than **integers**, which are encoded differently than **floating point numbers**, which are encoded differently than character **strings**.  The nice thing about using a higher level language like Python is that it works with the different data encodings at a level mostly invisible to the programmer.   

We will use the **assignment statements** to put data into memory.  Once the data is in memory we can manipulate the data using **expressions**.  We will soon learn about **arithmetic expressions** on numeric data and **string expressions** on string data.  When we use an expression to manipulate our data we will use an assignment statement to put the result of the expression in memory.  Once we have computed an expression and stored the result we will use the **print** function to display the result on a screen for us to see. 

The types of data we will cover here are **string** data, and two of the numeric data types: **integers**, and **floats**.

# Strings

A **string** in Python is a sequence of printable characters.  In Python we represent a string by putting the characters in quotes like the string below.
```python
"Hello World"
```
We call a character string in quotes a **string literal**.  When Python encounters a string literal in our Python code it converts the literal into a specific internal arrangement of 0's and 1's so it can be stored in computer memory.  When Python prints the string to some output device, say your monitor, the string of 0's and 1's in memory are converted to character codes for your monitor.  

> **Try It**<br>
1. Click your mouse cursor inside the code cell bellow and then type the literal `"Hello World"` 
2. Press the `Ctrl` and `Enter` key at the same time on the keyboard to execute the code cell.  Note in the future we refer to this combination as `Ctrl-Enter`. 

In [2]:
"Hello World"

'Hello World'

**What just happened?**

You entered the string "Hello World" into the code cell and execute it.  During execution the literal was echoed back on the line below called `Out[##]`.  

>**Experiment** <br><br>Now is a good time to experiment. Try using the `+` menu key to build a new cell below and then try the whole process again without the ending quote.  You will get an error.  Python follows strict syntax rules for what counts as code and if you break a rule Python doesn't know what to do.  When this happens Python stops trying to execute your code and displays an error message.  You should start getting use to the error messages Python uses and see what they mean since you will spend a lot of time fixing your code.  Most of the time `syntax errors` are easy to fix.          

## The Assignment Statement 

In the code cell below we create a variable named `greeting` and use the assignnment operator, `=`, to map the string literal `"Hello"` to greeting.  Now when we want to refer to the data `"Hello"` we will use the variable named greeting.

In [3]:
greeting = "Hello"
print(greeting)

Hello


**What just happened?**

* In line 1 build an assignment statement using the `=` operator in Python.
    * The Python interpreter notices the literal and places the string `Hello` into memory.
    * To the left of the `=` operator we typed the variable name `greeting`.
    * The Python interpreter will now know the variable named `greeting` points to the string `Hello` in memory.  
* In line 2 we use the variable `greeting` as an argument to the `print` function.
    * The Python interpreter then gets the data pointed to by the variable `greeting` and prints out the data in the space below the cell.

We can use the assignment operator to have a second variable name point to the same data in memory.  Execute the example below.  

> **Note:**<br>
In a notebook when you execute a cell it remembers the consequence of previous cell executions.  Thus we can say `b = greeting` below because we executed the code cell above.  If we fail to execute the cell above we will get an error when we try to execute `b = greeting`.

In [4]:
b = greeting
print(b)

Hello


Notice b also points to `Hello` in memory.  Now things will get a little complicated, but try to following whats happening in the example below.

In [5]:
greeting = "Hello Visitor"
b = greeting
print(greeting)
print(b)
print()

greeting = "Hello to you"
print(greeting)
print(b)
print()

b = greeting
print(b)

Hello Visitor
Hello Visitor

Hello to you
Hello Visitor

Hello to you


**What just happened?**

* In line 1 `greeting` points to `Hello Visitor` in memory.
* In line 2 `b` also points to the same `Hello Visitor` in memory.
* In line 7 `greeting` now points to `Hello to you` in memory, but `b` still points to `Hello Visitor`
* In line 12 `b` now point to the same string, i.e., `Hello to you`, in memory as the variable named `greeting`.

So after line 12 what happened to `Hello Visitor` in memory?  The answer is it may or may not still be there, but our program has no way to recall it since it is no longer associated with a variable name.


## Exercise  

Find the error in each code cell below and fix the code to make it work.  Notice the third one is a little tricky so look at the code closely.

In [6]:
greeting = "Hello World"
print(greeting)

Hello World


In [7]:
greetings = "Hello World"
print(greeting)

Hello World


In [8]:
greet = "Hello World"
print(greet)

Hello World


## f-strings

We can use f-strings to format strings.  Here is an example.

In [9]:
first_name = "Kevin"
specific_greeting = f"Hello {first_name}"  # Notice ths space between Hello and {first_name}
print(specific_greeting)

Hello Kevin


What just happened?

* In line one we assigned the string 'Kevin' to the variable `first_name`.
* In line two we used an f-string to format our specific_greeting string.
    * the letter `f""` followed by quotes denotes an f-string 
    * Inside the quotes we use braces `{}` to enclose the variable `first_name`.
    * The string in memory pointed to by `first_name` is substituted for the braces. 
* In line 2 we assign the memory contents of the f-string to the variable named specific_greeting. 
* In line 3 we now print the resulting string pointed to by specific_greeting.

# print()

We will write Python in Code cells in our notebook. When we are done entering code we will press `Ctrl-Enter` to run the code. Often the code will already be in the cell and all you have to do is press `Ctrl-Enter` to run it. Do this in the code cell below. 

In [10]:
# The first thing a programmer will write in a new langage is the
# ubiquitous "Hello World" program.
print("Hello World")
# Below this block of code you will see the result of the print statement.

Hello World


**What just happened?**

* **Lines 1, 2, and 4** above are comments.  Notice a Python comment starts with the hash symbol, #.  Python ignores comments but you can use them to explain what your code is doing.
* **Line 3** above is the first line of executable code.  It starts with the word `print` followed by the character string `"Hello World"` enclosed in left and right parentheses.  
    * When the cell above is executed (by pressing Ctrl-Enter inside that cell) the string literal is printed as`Hello World` without quotes to the terminal just below the code cell.
    * We call `print` a function and the literal `"Hello World"` an argument of the print function, and we note that arguments of a function are placed within left and right parentheses.
    * The print function a *built-in* function since it comes pre-installed with Python.   

## Basic Editing Practice 
You can edit code in a notebook in a way similar to using a word processor.  For example you can highlight code save it to a clip-board and then copy the code to another location.  You can also insert code and delete it.  Try this now.
>**Practice**
 1. Click on this cell.
 2. Make a new code cell below using the `+` menu item.  
 3. Highlight the `print("Hello World")` line in the code cell above using your mouse and press `Ctrl-c` to copy it.
 4. Click in the code cell you made below and press `Ctrl-v` to paste what you copied in the cell.
 5. Add something like "Hello World! I am here." By clicking your mouse after `World` and typing `! I am here.` just like a word processor.
6. Press Ctrl-Enter to run the cell.

In [11]:
print("Hello World! I am here.")

Hello World! I am here.


## Making Intentional Errors  
By purposefully making errors you learn what to expect what messages Python will produce when an error is encountered.  Later on when you make an unintentional error you have a better chance of fixing it.  try the following experiment below and notice the different types of error messages you get.

>**Experiment**<br>
 1. Make a new code cell below.  
 2. Type in `print("hello world"`  
     Without a closing right parentheses.
 3. Press `Ctrl-Enter` to run the cell.
 4. Next try `print(hello world)`  Notice, there are no quotes around `hello world`. 
 5. Also try `prin("hello world")`  Notice, print is misspelled as prin.

In [12]:
prin("hello world")

NameError: name 'prin' is not defined

## More on the print function

You can use the `end=""` argument to tell the print function what you want to to an the end of the print statement, and you can use the `sep=""` argument to tell the print function what to do when it finishes printing an argument in the print statement.  Here are some examples.

In [ ]:
# printing on a new line each time
print ("Hello")
print ("to")
print ("you")

# printing on the same line each time
print ("Hello", end = ' ')
print ("to", end = ' ')
print ("you")

# notice the space in the end = ' ' let's replace it with a hyphen.
print ("Hello", end = '-')
print ("to", end = '-')
print ("you")

# lets look at what sep("string") does
print ("Hello", "to", "you")  # notice multiple arguments are seperated by commas
print ("Hello", "to", "you", sep = '::')
print ("Hello", "to", "you", sep = '::', end = '!')

Hello
to
you
Hello to you
Hello-to-you
Hello to you
Hello::to::you
Hello::to::you!

>**Practice**<br>
 1. Click on this cell.
 2. Make a new code cell below using the `+` menu item.  
 3. See what happens if you type `print("")` and execute it. <br>
Notice the string is called the **empty string**, that is, nothing between the quotes.
 4. See what happens if you type `print()` and execute it.

In [ ]:
print()

# input()
So far we needed a number or string to use we entered it into our program.  A more general approach would be to let a person using our program enter the values they need when they run the program.  To do this, Python has a built in functions called `input`.  Here is an example. 

In [ ]:
first_name = input("Please enter your first name: ")
print(f"{greeting} {first_name}")

Please enter your first name:  Jon


Hello World Jon


What just happened?

* In **line one**, the `input` function asks the user for an input. 
* The wording of the request is the string argument `"Please enter your first name: "`. 
* When line one is run the request is printed followed by an input box.
* The program then waits for user input.    
* The input is converted to a string and the variable `first_name` is assigned the input string. 
* In **line two**, the `print` function is then called using an f-string as an argument.

# Numbers

We start by looking at two data types, integers, and floats. Some examples of this are shown in the next cell. Click on the next code cell and be sure to press Ctrl-Enter.

In [ ]:
# Quantities and price
quantity_apples = 10         
price_apple = 0.75          

What just happened?

We used two **assignment statements** to assign a variable name to a numeric value.  The equal sign,` =`, is used to express this assignment.  Whne we execute the assignment statement (with Ctrl-Enter) we place the numeric value in computer memory.  We can thenrefer to this value using the variable name.  The variable names are `quantity_apples` and `price_apple` while the values are `10` and `.75` respectively.  The value `10` is a whole number and is called an **integer**.  The value `.75` is a decimal number and is called a **float**.

# type(), str(), int(), float()

Python also has built in functions to convert the values of a variable to a different type.  For example we can use 

```python
str(var_name)
```
to convert the value assigned to `var_name` to a string.  We can also use

```python
type(var_name)
```
to determine the type of value assigned to `var_name`.  See more examples in the code below.

In [ ]:
a = 10
print(f"The type of value assigned to variable a is {type(a)}")

The type of value assigned to variable a is <class 'int'>


What just happened?

In **line two**, we use the`type` function in a `f-string` to `print` the `type` of value assigned to variable `a`. Notice the type of value in `a` is an integer as reported in `<class 'int'>`.  For now we will just not the `'int'`.  Later on we will talk about what `class` means.  

In [ ]:
a = str(a)
print(f"The type of value assigned to variable a is {type(a)}")

The type of value assigned to variable a is <class 'str'>


What just happened?

In **line one**, we used the `str` function to convert the value assigned to `a` to a string.
In **line two**, we use the`type` function in a `f-string` to `print` the `type` of value assigned to variable `a`. Notice the type of value in `a` is now reported as a sting, i.e., `<class 'str'>`. 

Here are some more conversions.  Look carefully at what each one does.

In [ ]:
a = 10
print(f"The value of a is {a} and its type is {type(a)}")
a = float(a)
print(f"The value of a is {a} and its type is {type(a)}")

The value of a is 10 and its type is <class 'int'>
The value of a is 10.0 and its type is <class 'float'>


In [ ]:
a = 7.7
print(f"The value of a is {a} and its type is {type(a)}")
a = int(a)
print(f"The value of a is {a} and its type is {type(a)}")

The value of a is 7.7 and its type is <class 'float'>
The value of a is 7 and its type is <class 'int'>


In [ ]:
a = 7.7
print(f"The value of a is {a} and its type is {type(a)}")
a = str(a)
print(f"The value of a is '{a}' and its type is {type(a)}")

The value of a is 7.7 and its type is <class 'float'>
The value of a is '7.7' and its type is <class 'str'>


# Arithmetic Expressions

We manipulate numeric data with arithmetic expressions which consist of **operators** together with **variable names**,  **constants**, and **functions**.

In [ ]:
quantity_apples = 10         
price_apple = 0.75  
apple_expense = quantity_apples * price_apple
print(apple_expense)

7.5


What just happened?

1. In **lines one and two**, we used assignment statements to give values to `quantity_apples` and `price_apple`.
2. In **line three**, we used an arithmetic expression to multiply `quantity_apples` and `price_apple` and assigned the resulting value to `apple_expense`.  Notice that multiplication is indicated by the operator `*`. 

>**Note:**<br><br>
You can also print numbers with the `print()` function.  In this case the Python interpreter converts the number to a string representation of the number and then prints the string.

The operators we can use in an arithmentic expression are the following:

            Operation     Symbol   Order

        * Parentheses       ()       1
        * Exponetiation     **       2
        * Unary Negation    -        3
        * Division          /        4
        * Multiplication    *        4
        * Modulus           %        4
        * Floor Division   //        4
        * Addition          +        5
        * Subtraction       -        5
        
Here are some examples of expressions.  Run the cells to see the results.  The number at the right indicate the order of precedence explained below.  To evaluate an expression the interpreter uses the order of precedence seen in the numbers above.  Lower numbers have precedence over higher number. Equal number (precedence) operations are evaluated from left to right. Note the () indicates an ordering operation and has the highest order of precedence.  Below is an example.    

In [ ]:
x = 6

print(f"the value of x**2 is {x**2}")
print(f"the value of -x**2 is {-x**2}")  #Note unary negation is applied after exponetiation.

#  To change the order of evaluation, use parentheses.
print(f"the value of (-x)**2 is {(-x)**2}")  # Now -x is done first


the value of x**2 is 36
the value of -x**2 is -36
the value of (-x)**2 is 36


>**Note:**<br><br>
You can also print numeric expressions in Python, or put them in f-strings like we did above.  The Python evaluates the numeric expression to a number which it then converts to a string. 

In [ ]:
x = 6
y = 3

print(f"the value of x/y*6 is {x/y*6}")

#  Division and Multiplication are evaluated left to right
#  To change the order of evaluation use parentheses.

print(f"the value of x/(y*6) is {x/(y*6)}")

the value of x/y*6 is 12.0
the value of x/(y*6) is 0.3333333333333333


**What just happened?**

When Python evaluates an expression it uses operator precedence to determine how to evaluate an expression.  In the example above division, /, and multiplication, +, have the same operator precedence.  When this occurs, e.g., in the expression `x/y*6`, Python evaluates from left to right, e.g., `6/3 = 2 and 2*6 = 12`.  You can change this order with parentheses, e.g., `x/(y*6)`.  Parentheses tell Python to evaluate what is inside the parentheses first, e.g., `3*6 = 18 and 6/18 = .3333333333333333` which is approximately 1/3.  Note, it is not quite 1/3 since a float can only have a finite number of decimal places and cannot express a number with infinite decimal places. 

In [ ]:
x = 6
y = 3
z = .5

print(f"the value of x/y*16**.5 is {x/y*6**.5}")

# in evaluating the expression x/y*6**.5
# the operations are done in this order
a = 6**.5  # First, exponentiation
print(a)
b = x/y    # Second, division (note now we are moving left to right)
print(b)
c = a*b    # Third, multiplication.
print(c)

the value of x/y*16**.5 is 4.898979485566356
2.449489742783178
2.0
4.898979485566356


### A.  Using a notbook in calculator mode

We can use a notebook cell in calculator mode as follows

In [ ]:
100**.5*25

250.0

Note, 100**.5 is the sqaure root of 100 or 10, and 10*25 = 250.  When a cell ends this way in an expression the result is simply printed out.  Notice, only the last expression is printed out.    

In [ ]:
100**.5*25
100*2

200

A variable name counts as a very simple expression.  Thus we can also do the following.

In [ ]:
y = 100**.5*25
y

250.0

### B.  Some unusual arithemtic operators

Python has two aritmetic operators, `%` or modulus, which returns the integer remainder of a division, and `//` or floor division, which returns an integer result of A division rounded down to the closest whole number.  Lets look at some examples below. 

In [ ]:
20/3

6.666666666666667

So this result is approximately 6 and 2/3.
Notice 20 - (6*3) = 2, or the remainder after whole division. 

In [ ]:
20 - int((20/3))*3

2

In [ ]:
20%3

2

Similarly, for floor division.

In [ ]:
int(20/3)

6

In [ ]:
20//3

6

In [ ]:
# here is a little program

numerator = int(input("Enter numerator: "))
denomenator = int(input("Enter denomeator: "))
print(f"{numerator}/{denomenator} = {numerator/denomenator}")
print(f"{numerator}%{denomenator} = {numerator%denomenator}")
print(f"{numerator}//{denomenator} = {numerator//denomenator}")
print(f"{numerator}/{denomenator} = {numerator//denomenator} with remainder {numerator%denomenator}")
print(f"{numerator}/{denomenator} = {numerator//denomenator} and {numerator%denomenator}/{denomenator}")

Enter numerator:  10
Enter denomeator:  5


10/5 = 2.0
10%5 = 0
10//5 = 2
10/5 = 2 with remainder 0
10/5 = 2 and 0/5


# Programming Style

The Python community uses **naming conventions** to help other programmers understand your code.  In Python, variable names should always be lowercase. Variable names with more than one word should be joined by an underscore. Finally, variable names should be chosen to describe the data values assigned to the name. As a rule don't make varaible names too long or they become hard to work with.

Comments that start with `#` are also used to help explain the intention behind the code.  A good programmer will choose variable names that are easy to understand in context and usually do not need additional comments.

There are many other style conventions that will help the later you and other programmers read and understand your code.  They can be found [here](https://www.python.org/dev/peps/pep-0008/)

In [ ]:
# Bad example of code
a = 25 #quantiy of apples
p = .25 #price per apples
# Now we will calculate the cost of apples
c = a * p
# Now we will print out the cost of apples
print(c)

6.25


In [ ]:
# Better Code

quantity_apples = 25
price_apple = .25
cost_apples = quantity_apples * price_apple
print(f"The cost of {quantity_apples} apples is {cost_apples}.")

The cost of 25 apples is 6.25.


# Example Program

We will now write a program that does temperature conversion from Celsius to Fahrenheit.  

In [ ]:
# Program to convert celsius to fahrenheit using the formula
# fahrenheit = 9/5 * Cclsius + 32
celsius = float(input("Enter temperature in Celsius: "))
fahrenheit = (9.0/5.0)*celsius + 32.0
print (f"{celsius}C = {fahrenheit}F")

Enter temperature in Celsius:  25


25.0C = 77.0F


# Exercise

Given a quadratic equation of the form,

$0=ax^2+bx+c$

Then the roots of this equations are given by

$x=\frac{-b \pm \sqrt{b^2-4ac}}{2a}$ 

In the code code cell below and calculate the roots for
a = 1, b = 3, c = -4.

In [7]:
# Use quadratic formula to solve for the rroots of ax^2 + bx + c 

a = 1   # quadratic coefficient
b = 3   # linear coefficient
c = -4  # constant

#TODO Use quadratic formula to find roots.

x = (-b + (b**2 - 4*a*c)**0.5)/(2*a)
print(f"The rroot of ax^2 + bx + c is {x}.")

The rroot of ax^2 + bx + c is 1.0.


But we know that the square root operation can sometimes produce a complex number, i.e., $\sqrt{-1}$.  Lucky for us python allows us to also use complex numbers.  Copy the code above to the code cell below but now change, a = 3.0, b = 4.0, c = 2.0.  See what you get.

In [8]:
# Use quadratic formula to solve for the rroots of ax^2 + bx + c 

a = 3.0
b = 4.0
c = 2.0

#TODO Copy your quadratic formula here.
x = (-b + (b**2 - 4*a*c)**0.5)/(2*a)
print(f"The rroot of ax^2 + bx + c is {x}.")


The rroot of ax^2 + bx + c is (-0.6666666666666666+0.47140452079103173j).


In the last step copy and change your code below to accept user input for a, b, and c.

In [9]:
# Use quadratic formula to solve for the rroots of ax^2 + bx + c 
# given user input for a, b, and c.

#TODO: Enter your code here
a = float(input("Enter a: "))
b = float(input("Enter b: "))
c = float(input("Enter c: "))
x = (-b + (b**2 - 4*a*c)**0.5)/(2*a)
print(f"The rroot of ax^2 + bx + c is {x}.")



Enter a:  20
Enter b:  10
Enter c:  10


The rroot of ax^2 + bx + c is (-0.24999999999999994+0.6614378277661477j).
